### PR TITLE
ARROW-13620: [R] Binding for n_distinct()

### PR DESCRIPTION
**Pull request by thisisnic**

### CODE REVIEW

**Review by nealrichardson — COMMENTED** [review]
One question; also can you please run the styler?

**Review by nealrichardson — APPROVED** [review]

**Comment by thisisnic** [comment]
@github-actions cross rebase

**Comment by thisisnic** [comment]
@github-actions rebase

File: r/tests/testthat/test-dplyr-aggregate.R [hunk 1]
```diff
@@ -235,6 +235,24 @@ test_that("Group by any/all", {
...
 
+test_that("Group by n_distinct() on dataset", {
+  expect_dplyr_equal(
+    input %>%
+      group_by(some_grouping) %>%
+      summarize(distinct = n_distinct(lgl, na.rm = FALSE)) %>%
+      collect(),
+    tbl
+  )
+  skip("ARROW-13764 - CountOptions (na.rm) not yet implemented for compute_distinct")
```
Comment (ianmcook) [inline r/tests/testthat/test-dplyr-aggregate.R:246]: Looks like ARROW-13764 is merged now
```suggestion
```

File: r/R/dplyr-functions.R [hunk 2]
```diff
@@ -825,6 +824,17 @@ agg_funcs$var <- function(x, na.rm = FALSE, ddof = 1) {
     options = list(ddof = ddof)
   )
 }
+
+agg_funcs$n_distinct <- function(x, na.rm = FALSE) {
+  list(
+    fun = "count_distinct",
+    data = x,
+    # ARROW-13764 Passing in na.rm = TRUE doesn't actually work yet as
+    # CountOptions not yet implemented for count_distinct
```
Comment (ianmcook) [inline r/R/dplyr-functions.R:833]:
```suggestion
```

Comment (thisisnic) [reply r/R/dplyr-functions.R:833]: Thanks!

**Pull request closed without merging**